### PR TITLE
daemon: Introduce the minimal fx application

### PR DESCRIPTION
**Pull request by joamaki**
As the first step towards a more modular cilium-agent
application, bring in uber/fx by wrapping runDaemon
and lifting *Daemon into the dependency graph.

This retains all existing behaviour, but will allow us
to incrementally lift modules out from Daemon.

### CODE REVIEW

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: Small nit here: Per https://go.dev/tour/flowcontrol/13 , the first defer that is added is the last defer to be run, and vice versa. In practice this means that the `srv.Shutdown()` below is run first, presumably then closing the API to stop receiving events from the API outside (eg endpoint creation/deletion and so on), then later on the Daemon is unsubscribed from the Daemon, thereby unhooking the IPAM & monitor event processors. From what I'm reading, it seems like the `Add()` here is actually switching the order around, so technically the API could still be live but the agent won't internally handle the event. I think it makes sense to retain the current order if we can? This is a bit of a race condition only during shutdown, but if it's not a huge effort then it probably makes sense to retain this in the desired order.

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go]: :thinking: usually the place that creates the context&cancel will be the one to initiate the cancel. Is it correct for this function to now grab control over the cancel function? For that matter, passing the cancel function down is a little unusual in itself.

Comment (joestringer) [reply daemon/cmd/daemon_main.go]: ^ I didn't think too hard about this yet, but maybe this function should be enlightened to return a `fx.Option` and hence use `fx.Error()` in cases like this?

The comparable thing in https://pkg.go.dev/go.uber.org/fx#Error would be to do this, and then also for the return result of this function at the end, wrap the returned `*Daemon` in an `fx.Provide(...)` call.

File: daemon/cmd/app.go
Comment (joestringer) [inline daemon/cmd/app.go]: Ah, now I see why the current implementation passes the `cancelFunc` down. I guess this works for now, though a bit odd. Maybe we can explore using [App.Done()](https://pkg.go.dev/go.uber.org/fx#App.Done) and `Shutdowner`s to do this more in line with the way that fx expects. Minor nits though, it's not super clear to me whether there's a motivation to look into this specifically now as part of this PR or consider handling it in future. May partly depend on where we think the code is eventually going to get to.

File: daemon/cmd/app.go
Comment (joestringer) [inline daemon/cmd/app.go]: If you grep around for `time` in `pkg/defaults/defaults.go`, you'll see some other similar startup timeouts, which I think are all configurable. This sort-of fits with them, but for instance if there are a lot of k8s objects to sync then it may make sense to bump this timeout... Maybe we can add an option for each of these to allow users to override the timeouts just in case?

File: daemon/cmd/app.go
Comment (joestringer) [inline daemon/cmd/app.go]: nit:
```suggestion
// daemonModule wraps the existing daemon initialization
```

Even if we intend for this to be temporary, the reality is that after this PR is merged, there is a daemon module. We can argue about whether it should be there or not, but until we make enough changes to actually remove the need for an overall daemon module, it's going to stay here. And as always, even with the best intentions, code you think is temporary will likely live longer than you expect.

File: daemon/cmd/app_logger.go
Comment (joestringer) [inline daemon/cmd/app_logger.go]: Is this deliberately exposing these logs as a different subsystem from `daemonSubsys`? So far I've mostly assumed that a "subsystem" is for the most part a specific directory. For all the daemon stuff we just use `daemon` but for all the various `pkg/foo/bar` we typically name these with package-level names like `foo-bar` matching the path. This way, when there's a log, you can clearly see which package the log came from by looking at the subsys field.
```suggestion
		logging.DefaultLogger.WithField(logfields.LogSubsys, daemonSubsys),
```

**Review by joestringer — DISMISSED** [review]
Mostly just minor nits, there's one functional change that I observed here in the ordering of the shutdown operations. Otherwise LGTM as a first step.

File: daemon/cmd/app_logger.go
Comment (joamaki) [reply daemon/cmd/app_logger.go]: Yes I was being deliberate in doing that in order to be able to find logs related to fx itself, but your reasoning with sticking to package names makes sense. Perhaps there's another field we could use for this?

File: daemon/cmd/app.go
Comment (joamaki) [reply daemon/cmd/app.go]: Good point. I'm going to go with the suggestion you had that `runDaemon` returns an fx.Option, which removes the need for this and will leave less stuff that sticks around :)

File: daemon/cmd/app.go
Comment (joamaki) [reply daemon/cmd/app.go]: Currently there's no start hooks so the timeout does nothing. Perhaps the right thing to do here is just use that one context we have in daemon for both start and stop and then revisit when start and stop actually becomes meaningful?

EDIT: Removed these constants and made the start use the daemon context, which is cancelled on signals, and made the stop not have a timeout for now. I think this is better than using something arbitrary.

File: daemon/cmd/app.go
Comment (joamaki) [reply daemon/cmd/app.go]: Yep, it's a bit unsatisfying how this tries to stick with the existing way of stopping things. My take on this currently is to not yet touch these parts, but perhaps as one of the first follow ups refactor the signal handling and cleanup to work the fx way? EDIT: Tried making "daemonCleanup" a module and it ended up being straightforward, so that'll be the follow-up PR after this one.

File: daemon/cmd/daemon_main.go
Comment (joamaki) [reply daemon/cmd/daemon_main.go]: That's actually a good idea and a bit cleaner. I'll try it out. Will likely still need to keep a call to cancel() for example for the case where the `srv.Serve()` would return. But as I said in another comment, lets try to do the minimal change now, and do a follow-up that focuses just on fx'ifying the cleanup.

File: daemon/cmd/daemon_main.go
Comment (joamaki) [reply daemon/cmd/daemon_main.go]: Oh wow and here I was being confident that now at least I'm definitely not doing any functional change. This if anything highlights that we'll need to move in very small increments and review these very carefully.

I'll defer the add to retain the current order and add a comment about it since the defer can be a bit puzzling.

File: daemon/cmd/app.go
Comment (joamaki) [reply daemon/cmd/app.go]: I included this now into this PR as the changes were pretty minimal.

File: daemon/cmd/daemon_main.go
Comment (tklauser) [inline daemon/cmd/daemon_main.go]: It looks like gops agent integration is removed entirely by this PR. Is this intentional? I think we preferably want to keep gops for debugging.

Comment (joamaki) [reply daemon/cmd/daemon_main.go]: Oops, definitely was not intentional! Rebasing accident. I'll fix asap.

**Review by tklauser — APPROVED** [review]
:rocket:

**Review by pippolo84 — APPROVED** [review]
LGTM :rocket:

**Review by pippolo84 — APPROVED** [review]

**Comment by pchaigno** [comment]
Do we want to wait for the v1.12 release before merging this? Seems like it will likely cause a lot of backporting conflicts :-/

**Comment by joamaki** [comment]
> Do we want to wait for the v1.12 release before merging this? Seems like it will likely cause a lot of backporting conflicts :-/

I took a look at all the backport commits in v1.11 and there's very few that touch code in `daemon/cmd`. Mostly flags were added and none of them conflict with changes in this PR. I'm fairly certain this PR won't cause big conflicts.

The follow-ups to this PR are of course another matter and with those I'm happy to wait. The reason why I'd want this in soonish is that once this is merged I can do follow-ups in parallel.

File: daemon/cmd/app_logger.go
Comment (aanm) [inline daemon/cmd/app_logger.go:21]: Why do we need to log all events? Is it helpful to have these printed while we are transition to uber/fx?

File: daemon/cmd/app_logger.go
Comment (aanm) [inline daemon/cmd/app_logger.go:4]: needs copyright header

File: daemon/cmd/app.go
Comment (aanm) [inline daemon/cmd/app.go:19]: Do we need this line?

File: daemon/cmd/cleanup.go
Comment (aanm) [inline daemon/cmd/cleanup.go:64]: Where did we have the signal handlers?

**Review by aanm — COMMENTED** [review]
This looks suspiciously simple :thinking:. Overall LGTM I just have a couple questions

File: daemon/cmd/app_logger.go
Comment (joamaki) [reply daemon/cmd/app_logger.go:21]: There's not that many of them, and they're very helpful as they show how long things took to initialize and start. Most of them are `Debug()` so I think the only thing that really shows up by default is "Started" when everything has finished starting up.

Here's what it looks like with debugging on (in a branch with more modules): 
```
level=debug msg="OnStart hook executing" callee="github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).onStart-fm()" caller=github.com/cilium/cilium/pkg/endpointmanager.newEndpointManager subsys=daemon
level=debug msg="Starting new controller" name=endpoint-gc subsys=controller uuid=9bb71dac-f984-4ceb-b6e7-27b3f6e03426
level=debug msg="OnStart hook executed" callee="github.com/cilium/cilium/pkg/endpointmanager.(*endpointManager).onStart-fm()" caller=github.com/cilium/cilium/pkg/endpointmanager.newEndpointManager runtime="36.159µs" subsys=daemon
level=debug msg="OnStart hook executing" callee="github.com/cilium/cilium/pkg/status.(*collector).start-fm()" caller=github.com/cilium/cilium/pkg/status.newCollector subsys=daemon
level=debug msg="OnStart hook executed" callee="github.com/cilium/cilium/pkg/status.(*collector).start-fm()" caller=github.com/cilium/cilium/pkg/status.newCollector runtime="4.358µs" subsys=daemon
level=debug msg="OnStart hook executing" callee="github.com/cilium/cilium/api/v1/server.(*Server).startHook-fm()" caller=github.com/cilium/cilium/api/v1/server.newServer subsys=daemon
level=debug msg="OnStart hook executed" callee="github.com/cilium/cilium/api/v1/server.(*Server).startHook-fm()" caller=github.com/cilium/cilium/api/v1/server.newServer runtime="1.102µs" subsys=daemon
level=debug msg="OnStart hook executing" callee="github.com/cilium/cilium/daemon/cmd.gops.start-fm()" caller=github.com/cilium/cilium/daemon/cmd.registerGops subsys=daemon
```

File: daemon/cmd/app.go
Comment (joamaki) [reply daemon/cmd/app.go:19]: Yes, for now as nothing depends on daemon and we need an invoke to pull on the constructors. This will go away with the next PR.

File: daemon/cmd/cleanup.go
Comment (joamaki) [reply daemon/cmd/cleanup.go:64]: We're now relying on uber/fx's signal handling (via `app.Run()`).

File: daemon/cmd/app_logger.go
Comment (joamaki) [reply daemon/cmd/app_logger.go:4]: Fixed.

File: daemon/cmd/cleanup.go
Comment (joamaki) [reply daemon/cmd/cleanup.go:64]: https://github.com/uber-go/fx/blob/4f51a19264fc8d0bf7b87ef7a8855feb96d264b7/app.go#L716

File: daemon/cmd/cleanup.go
Comment (joestringer) [inline daemon/cmd/cleanup.go:91]: It looks like we threw away the gops.Close() with the signal cleanup bits, which means it won't be invoked any more.

Incidentally, the way that gops is invoked today has bothered me for a while, since it doesn't take into account the commandline flags; it should probably be moved into the daemon and after the CLI flag parsing. Not sure quite how you'd like to handle this minimally for now, but open to suggestions. My minimal suggestion right now is just that I think `gops.Close()` should be invoked on shutdown somehow.

File: daemon/cmd/daemon_main.go
Comment (joestringer) [inline daemon/cmd/daemon_main.go:1668]: Hmm, there's something subtle going on here. Rather than deriving the context from the server (ie the API), this is creating the daemon with `context.Background()` via `runApp()`.

By the looks, the `api/v1/server/configure_cilium_api.go:ServerCtx` is just derived from `context.Background()` anyway but that one is cancelled via the Server Shutdown. So I guess we had this sorta backwards before? Now with this new code, the daemon/ code defines the original context, then based on the fx module handling it will cancel that context, then trigger the API shutdown via the new Cleaner module.

I think that's fine.

File: daemon/cmd/daemon_test.go
Comment (joestringer) [inline daemon/cmd/daemon_test.go:137]: When does this get cancelled? Does it need to be?

**Review by joestringer — CHANGES_REQUESTED** [review]
LGTM, it just looks like the gops shutdown was removed.

File: daemon/cmd/daemon_test.go
Comment (joamaki) [reply daemon/cmd/daemon_test.go:137]: Cancelled here as the first thing when the application is stopped: https://github.com/cilium/cilium/pull/19795/files#diff-85ea165d03636d6330e7059ac116133ee9cfedcb7ead9690d3289019a9698b5aR23.

File: daemon/cmd/daemon_main.go
Comment (joamaki) [reply daemon/cmd/daemon_main.go:1668]: Yep, earlier the `ServerCtx` was the root context and this now introduces a new root context for the application. I'll follow with a PR soon which cleans this further by making the API server a module with start and stop hooks.

File: daemon/cmd/cleanup.go
Comment (joamaki) [reply daemon/cmd/cleanup.go:91]: The refactoring that screwed this up was making a gops into a module and that will be then started after flag parsing. I'll follow with a PR to do that soon. 

I'll add the missing `gops.Close` here though!

File: daemon/cmd/daemon_test.go
Comment (joestringer) [reply daemon/cmd/daemon_test.go:137]: :thinking: the line you linked refers to a cancel function declared 10 lines above, in the same function as the link. That function is different from the one I commented on here. The cancel gets stored in the test function version of the structure, so probably not getting cancelled now..?

Comment (joamaki) [reply daemon/cmd/daemon_test.go:137]: Doh, looks like I can't read today. Yep, indeed it's not cancelled in the tests. I'll add it.

**Review by joestringer — APPROVED** [review]

**Pull request merged**
